### PR TITLE
Refactor payload

### PR DESCRIPTION
**Pull request by keto**

### CODE REVIEW

File: src/webhook_launcher/app/payload.py [hunk 1]
```diff
@@ -1,124 +1,78 @@
...
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 # GNU General Public License for more details.
-# 
+#
 # You should have received a copy of the GNU General Public License
 # along with this program; if not, write to the Free Software
-# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+# Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
```
Comment (larstiq) [inline src/webhook_launcher/app/payload.py]: Cosmetic: Perhaps but some of this on line 16? The address breakage looks weird.

File: src/webhook_launcher/app/payload.py [hunk 2]
```diff
@@ -1,124 +1,78 @@
...
+from __future__ import print_function
```
Comment (larstiq) [inline src/webhook_launcher/app/payload.py]: Is this to make it able to run on python3? I'm still python2 enough that print_function is not in my muscle memory.

File: src/webhook_launcher/app/payload.py [hunk 3]
```diff
@@ -1,124 +1,78 @@
...
+    for klass in [
+        BbPush,
+        GhPush,
+        # Fallback to Payload base class which handles unknown formats
+        Payload,
+    ]:
+        try:
+            payload = klass(data)
             break
+        except:
```
Comment (larstiq) [inline src/webhook_launcher/app/payload.py]: Should this be except PayloadParsingError?

File: src/webhook_launcher/app/payload.py [hunk 4]
```diff
@@ -129,138 +83,162 @@ def create_placeholder(self, repourl, branch, packages=None):
...
-        official_packages = list(set(mapobj.package for mapobj in
-                                WebHookMapping.objects.filter(repourl=self.url,
-                                project__in=official_projects).exclude(package="")))
+        official_projects = list(
+            set(prj.name for prj in
+                Project.objects.filter(official=True, allowed=True)
+                )
+        )
+        official_packages = list(
+            set(mapobj.package for mapobj in
```
Comment (larstiq) [inline src/webhook_launcher/app/payload.py]: Cosmetic: I find this style where after the opening brace there is still stuff, `set(map`, but the opening brace on its own line a bit awkward. Add a newline and reflow?  Also the other occurences.

File: src/webhook_launcher/app/test/test_payload.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                {'name': 'master', 'changeset': 'otherrevision'}
+            ],
+            'tags': [
+                {'name': '1.0', 'changeset': 'otherrevision'}
+            ]
+        }
+        payload.data['commits'] = []
+        payload.handle()
+        api_mock.branches_tags.assert_called_once()
+        launch_build.assert_called_once()
```
Comment (larstiq) [inline src/webhook_launcher/app/test/test_payload.py]: Test gh build for consistency? And maybe split into two tests.

Comment (keto) [inline src/webhook_launcher/app/test/test_payload.py]: Added TODO about that, needs proper tag push payload

**Pull request merged**